Stop storing `Connection::SyncMessageState` as a raw pointer inside a HashMap

cdumez opened the pull request:
#### 18ce3a5fb18350ddef137659855acf961be40c22
<pre>
Stop storing `Connection::SyncMessageState` as a raw pointer inside a HashMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=302607">https://bugs.webkit.org/show_bug.cgi?id=302607</a>

Reviewed by Darin Adler.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::getOrCreate):
(IPC::Connection::SyncMessageState::~SyncMessageState):
(IPC::Connection::enqueueMatchingMessagesToMessageReceiveQueue):
(IPC::Connection::open):
(IPC::Connection::waitForMessage):
(IPC::Connection::waitForSyncReply):
(IPC::Connection::processIncomingSyncReply):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatcher):
(IPC::Connection::dispatchSyncStateMessages):
(IPC::Connection::protectedSyncState const):
(IPC::Connection::SyncMessageState::dispatcher): Deleted.
(IPC::Connection::SyncMessageState::wakeUpClientRunLoop): Deleted.
(IPC::Connection::SyncMessageState::wait): Deleted.
(IPC::Connection::SyncMessageState::SyncMessageState): Deleted.
(IPC::Connection::SyncMessageState::ConnectionAndIncomingMessage::dispatch): Deleted.
(IPC::Connection::SyncMessageState::WTF_GUARDED_BY_LOCK): Deleted.
(IPC::Connection::SyncMessageState::get):
(IPC::Connection::SyncMessageStateRelease::operator() const): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:

Canonical link: <a href="https://commits.webkit.org/303164@main">https://commits.webkit.org/303164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee7c214c153677804dd0a51268214735ad652a25

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139011 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a4df0c4f-7ab6-49d4-ba01-7a908e261964) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3860 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3676 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b7a098e4-1fcd-4d7d-b550-52a2dab6280a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134450 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117706 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81195 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ef2772d9-901c-4781-8f53-7c77a3f33f85) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/432 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82203 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141657 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3580 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36350 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/108766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3628 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3189 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/align_center_position_50.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108997 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27626 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2710 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114037 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56768 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3641 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32443 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3466 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67052 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3571 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -143,35 +143,27 @@ class Connection::SyncMessageState {
-std::unique_ptr<Connection::SyncMessageState, Connection::SyncMessageStateRelease> Connection::SyncMessageState::get(SerialFunctionDispatcher& dispatcher)
+Ref<Connection::SyncMessageState> Connection::SyncMessageState::getOrCreate(SerialFunctionDispatcher& dispatcher)
 {
+    RefPtr<SyncMessageState> newState;
```
Why is this local variable being defined outside the lambda? Is this perhaps left over from before some refactoring?

cdumez replied:
This looks needed to me? What else would keep the SyncMessageState alive when we have to create it? The HashMap itself only stores a `ThreadSafeWeakPtr<SyncMessageState>`. I think we need to keep it in scope until we end up calling `iterator.value.get()` to get a `RefPtr` from the `ThreadSafeWeakPtr`. Am I missing something?

darinadler replied:
I see your point, but it seems subtle, non-obvious to me. Maybe return `newState.releaseNonNull()` instead would make it more obvious?

cdumez replied:
Hmm, `newState` only gets initialized when we construct a new state though. It would be null in the case where the entry is already in the map. Maybe I can find a way to write this more clearly..

cdumez replied:
Will rewrite it this way which I'm hoping is clearer:
```
Ref<Connection::SyncMessageState> Connection::SyncMessageState::getOrCreate(SerialFunctionDispatcher& dispatcher)
{
    Locker locker { syncMessageStateMapLock };
    auto addResult = syncMessageStateMap().add(&dispatcher, nullptr);
    if (!addResult.isNewEntry)
        return addResult.iterator->value.get().releaseNonNull();
    Ref newState = adoptRef(*new SyncMessageState(dispatcher));
    addResult.iterator->value = newState.get();
    return newState;
}
```

darinadler commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -74,10 +74,11 @@ static UnfairLock s_connectionMapLock;
 public:
-    static std::unique_ptr<SyncMessageState, SyncMessageStateRelease> get(SerialFunctionDispatcher&);
-    RefPtr<SerialFunctionDispatcher> dispatcher() { return m_dispatcher.get(); }
+    static Ref<SyncMessageState> getOrCreate(SerialFunctionDispatcher&);
```
I find this name a bit clumsy. Don’t we use the word “ensure” for this elsewhere? Or is there some other single-word name that would do?

cdumez replied:
Oh, I can use `ensure`. Note that `getOrCreate` is quite common in our codebase too.

cdumez replied:
Personally, I think we use `bar.ensureFoo()` to get a Foo from Bar that may get lazy initialized. However, I think we use `Foo::getOrCreate()` when it is a factory function on Foo itself which may return an existing instance (thus getOrCreate instead of create). I have never seen `Foo::ensure()` and `Foo::ensureFoo()` would look odd I think.

darinadler replied:
OK, I accept that _ensure_ is not the right word. I would love us to find a single-word name for such functions, because, as I said, the name _get or create_ feels clumsy.

darinadler replied:
But if we don’t have an idea, no need to hold up this pull request.

darinadler commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -74,10 +74,11 @@ static UnfairLock s_connectionMapLock;
-    static std::unique_ptr<SyncMessageState, SyncMessageStateRelease> get(SerialFunctionDispatcher&);
-    RefPtr<SerialFunctionDispatcher> dispatcher() { return m_dispatcher.get(); }
+    static Ref<SyncMessageState> getOrCreate(SerialFunctionDispatcher&);
+    Ref<SerialFunctionDispatcher> dispatcher() { return m_dispatcher.get().releaseNonNull(); }
```
Should this be const?

cdumez replied:
Ok

darinadler commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1521,8 +1514,9 @@ SerialFunctionDispatcher& Connection::dispatcher()
-    RELEASE_ASSERT(m_syncState);
-    RefPtr dispatcher = m_syncState->dispatcher();
+    RefPtr syncState = m_syncState;
+    RELEASE_ASSERT(syncState);
+    RefPtr dispatcher = syncState->dispatcher();
```
Doesn’t the `RefPtr` `->` operator already do the equivalent of a `RELEASE_ASSERT`?

cdumez replied:
It doesn't. We tried but this regressed performance.

cdumez committed:
Stop storing `Connection::SyncMessageState` as a raw pointer inside a HashMap
https://bugs.webkit.org/show_bug.cgi?id=302607

Reviewed by Darin Adler.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::getOrCreate):
(IPC::Connection::SyncMessageState::~SyncMessageState):
(IPC::Connection::enqueueMatchingMessagesToMessageReceiveQueue):
(IPC::Connection::open):
(IPC::Connection::waitForMessage):
(IPC::Connection::waitForSyncReply):
(IPC::Connection::processIncomingSyncReply):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatcher):
(IPC::Connection::dispatchSyncStateMessages):
(IPC::Connection::protectedSyncState const):
(IPC::Connection::SyncMessageState::dispatcher): Deleted.
(IPC::Connection::SyncMessageState::wakeUpClientRunLoop): Deleted.
(IPC::Connection::SyncMessageState::wait): Deleted.
(IPC::Connection::SyncMessageState::SyncMessageState): Deleted.
(IPC::Connection::SyncMessageState::ConnectionAndIncomingMessage::dispatch): Deleted.
(IPC::Connection::SyncMessageState::WTF_GUARDED_BY_LOCK): Deleted.
(IPC::Connection::SyncMessageState::get):
(IPC::Connection::SyncMessageStateRelease::operator() const): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:

Canonical link: https://commits.webkit.org/303164@main
